metrics/measured_http/http.go

```diff
@@ -0,0 +1,91 @@
+package measured_http
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/jmhodges/clock"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	responseTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "response_time",
+			Help: "Time taken to respond to a request",
+		},
+		[]string{"endpoint", "method", "code"})
+)
+
+func init() {
+	prometheus.MustRegister(responseTime)
+}
+
+// responseWriterWithStatus satisfies http.ResponseWriter, but keeps track of the
+// status code for gathering stats.
+type responseWriterWithStatus struct {
+	http.ResponseWriter
+	code int
+}
+
+// WriteHeader stores a status code for generating stats.
+func (r *responseWriterWithStatus) WriteHeader(code int) {
+	r.code = code
+	r.ResponseWriter.WriteHeader(code)
+}
+
+// MeasuredHandler wraps an http.Handler and records prometheus stats
+type MeasuredHandler struct {
+	http.Handler
+	clk clock.Clock
+	// Normally this is always responseTime, but we override it for testing.
+	stat *prometheus.HistogramVec
+}
+
+func New(h http.Handler, clk clock.Clock) *MeasuredHandler {
+	return &MeasuredHandler{
+		Handler: h,
+		clk:     clk,
+		stat:    responseTime,
+	}
+}
+
+// endpointFromPath turns a request path into a value suitable for a Prometheus
+// label value, by eliminating path components that vary widely (like user ids
+// and authorization ids). It uses a simple heuristic: Remove everything after
+// the first path component that doesn't match lowercase a-z, plus hyphen. This
+// happens to work very well with Boulder's WFE.
```

Comment:
Should this heuristic be pulled out of this file and documented somewhere (in the WFE? In contributing.md?) so that we don't incidentally add something that breaks it? Seems unlikely - I'm ok either way, just want to bring this thought forward.